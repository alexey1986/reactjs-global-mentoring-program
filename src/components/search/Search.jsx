import React from "react";
import { PropTypes } from 'prop-types';
import { Col, Form, Button } from "react-bootstrap";
import styles from "./styles.module.css";
import { text } from "../../data.js";

const Search = ({ submitHandler, changeHandler, defaultValue }) => {
    const { placeholderTxt, findMovieTxt, searchTxt } = text;

    return (
        <Form onSubmit={submitHandler}>
            <Form.Label htmlFor="searchField" className={styles.formLabel}>
                {findMovieTxt}
            </Form.Label>
            <Form.Row>
                <Col>
                    <Form.Group controlId="searchField">
                        <Form.Control
                            name="searchMovie"
                            type="search"
                            defaultValue={defaultValue}
                            placeholder={placeholderTxt}
                            onChange={changeHandler}
                        />
                    </Form.Group>
                </Col>
                <Col>
                    <Button type="submit" >{searchTxt}</Button>
                </Col>
            </Form.Row>
        </Form>
    );
};

Search.propTypes = {
    submitHandler: PropTypes.func,
    changeHandler: PropTypes.func,
    defaultValue: PropTypes.string
};

export default Search;
