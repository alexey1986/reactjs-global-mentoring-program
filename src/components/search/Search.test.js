import React from "react";
import Search from "./Search";
import { render, screen, fireEvent, cleanup } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";

describe("Search:", () => {

    afterEach(cleanup);
    test("renders search component", () => {
        render(<Search />);
        expect(screen.getByText(/Find your movie/i)).toBeInTheDocument();
        expect(screen.getByText(/Search/i)).toBeInTheDocument();
        expect(screen.getByRole("searchbox")).toBeInTheDocument();
        expect(screen.getByRole("button")).toBeInTheDocument();
    });

    test("to matches snapshot", () => {
        const { asFragment } = render(<Search />);
        expect(asFragment(<Search />)).toMatchSnapshot();
    });

    test("search input should have a new value", () => {
        const { getByRole } = render(<Search />);
        const searchbox = getByRole(/searchbox/i);
        expect(searchbox).toHaveValue('');
        fireEvent.change(searchbox, { target: { value: "Jedi" } });
        expect(searchbox.value).toEqual("Jedi");
    });

    test("should call change event", () => {
        const { getByRole } = render(<Search />);
        const handleChange = jest.fn();
        const searchbox = getByRole(/searchbox/i);
        searchbox.onchange = handleChange;
        fireEvent.change(searchbox, { target: { value: 'Jedi' } });
        expect(handleChange).toHaveBeenCalled();
    });

    it('userEvent with callback provided as a prop', () => {
        const handleChange = jest.fn();
        const { getByRole } = render(<Search changeHandler={handleChange} />);
        const searchbox = getByRole(/searchbox/i);
        userEvent.type(searchbox, 'Mandalorian');
        expect(handleChange).toHaveBeenCalledTimes(11);
    });

    it('should submit form', async () => {
        const handleSubmit = jest.fn(e => e.preventDefault());
        const { getByRole } = render(<Search submitHandler={handleSubmit} />);
        const searchbox = getByRole(/searchbox/i);
        const button = getByRole(/button/i);
        fireEvent.change(searchbox, 'jedi');
        fireEvent.click(button);
        expect(handleSubmit).toHaveBeenCalled();
    });
});
