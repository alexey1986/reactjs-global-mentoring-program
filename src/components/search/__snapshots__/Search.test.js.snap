// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search: to matches snapshot 1`] = `
<DocumentFragment>
  <form
    class=""
  >
    <label
      class="formLabel form-label"
      for="searchField"
    >
      Find your movie
    </label>
    <div
      class="form-row"
    >
      <div
        class="col"
      >
        <div
          class="form-group"
        >
          <input
            class="form-control"
            id="searchField"
            name="searchMovie"
            placeholder="What do you want to watch?"
            type="search"
            value=""
          />
        </div>
      </div>
      <div
        class="col"
      >
        <button
          class="btn btn-primary"
          type="submit"
        >
          Search
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
